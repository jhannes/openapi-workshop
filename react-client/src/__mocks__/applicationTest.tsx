import * as React from "react";
import { ReactElement, ReactNode } from "react";
import { act, Simulate } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { ApplicationApis } from "@jhannes/openapi-workshop/dist/api";
import { activeDirectory } from "@jhannes/openapi-workshop";
import { mockApplicationApis } from "@jhannes/openapi-workshop/dist/test/apiTest";
import { ApiContext, identityProvider } from "../applicationContext";
import { MemoryRouter } from "react-router";
import {
  SampleModelFactories,
  TestSampleData,
} from "@jhannes/openapi-workshop/dist/test/modelTest";

export async function render(component: ReactElement) {
  const container = document.createElement("div");
  await act(async () => {
    ReactDOM.render(component, container);
  });
  return container;
}

export async function renderApplication(
  component: ReactNode,
  apis: Partial<ApplicationApis> = {},
  security: activeDirectory = new activeDirectory("")
) {
  const context = {
    apis: mockApplicationApis(apis),
    identityProvider,
    security,
  };
  return await render(
    <ApiContext.Provider value={context}>
      <MemoryRouter>{component}</MemoryRouter>
    </ApiContext.Provider>
  );
}

export function simulateFormChange(
  container: Element,
  label: string,
  value?: string
) {
  const element = container.querySelector(
    `[data-label=${label}] input, [data-label=${label}] select`
  );
  if (!element) {
    expect(`Could not find label[data-label=${label}]`).toBeUndefined();
    return;
  }
  const target: any = { value };
  Simulate.change(element, { target });
}

export function simulateCheck(
  container: Element,
  label: string,
  checked: boolean
) {
  const element = container.querySelector(
    `[data-label=${label}] input[type=checkbox]`
  );
  if (!element) {
    expect(`Could not find label[data-label=${label}]`).toBeUndefined();
    return;
  }
  const target: any = { checked };
  Simulate.change(element, { target });
}

export function testData(seed: number) {
  const sampleModelProperties: SampleModelFactories = {
    CategoryDto: {
      name: (sample) => sample.pickOne(["Cat", "Dog", "Bird", "Fish"]),
    },
    PetDto: {
      name: (sample) =>
        sample.pickOne(["Buddy", "Rocky", "Chase", "Daisy", "Tiger"]),
    },
  };
  return new TestSampleData({ seed, sampleModelProperties });
}
