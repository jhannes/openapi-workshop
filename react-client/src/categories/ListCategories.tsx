import * as React from "react";
import { useContext } from "react";
import { ApiContext } from "../applicationContext";
import { useLoader } from "../lib/useLoader";
import { LoadingView } from "../views/LoadingView";
import { ErrorView } from "../views/ErrorView";

export function ListCategories() {
  const {
    apis: { petApi },
  } = useContext(ApiContext);

  const categories = useLoader(async () => await petApi.listCategories());

  if (categories.loading) {
    return <LoadingView />;
  }
  if (categories.failed) {
    return <ErrorView error={categories.error} />;
  }
  return (
    <ul>
      {categories.data.map(({ id, name }) => (
        <li key={id}>{name}</li>
      ))}
    </ul>
  );
}
