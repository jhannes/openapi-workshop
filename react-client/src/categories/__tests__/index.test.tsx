import { renderApplication, testData } from "../../__mocks__/applicationTest";
import { ListCategories } from "../ListCategories";
import pretty from "pretty";
import React from "react";
import { mockPetApi } from "../../generated/test/apiTest";

describe("categories views", () => {
  it("shows categories", async () => {
    const sampleData = testData(300);
    const container = await renderApplication(<ListCategories />, {
      petApi: mockPetApi({
        listCategories: async () => sampleData.sampleArrayCategoryDto(),
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("shows errors", async () => {
    const container = await renderApplication(<ListCategories />, {
      petApi: mockPetApi({
        listCategories: async () => {
          throw new Error("Something went wrong");
        },
      }),
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
