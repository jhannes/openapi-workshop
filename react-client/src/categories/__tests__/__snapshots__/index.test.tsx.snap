// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categories views shows categories 1`] = `
"<ul>
  <li>Bird</li>
</ul>"
`;

exports[`categories views shows errors 1`] = `
"<div>
  <h1>An error has occurred</h1>
  <div>Error: Something went wrong</div>
</div>"
`;
